File: mobile/lib/voice_commands.dart
enum VoiceCommandKind {
  createClass,
  createRelationship,
  addAttribute,
  addMethod,
  undoVoiceCommand,
}

enum VoiceRelationshipType {
  association,
  aggregation,
  composition,
  generalization,
  realization,
  dependency,
}

class VoiceCommand {
  const VoiceCommand({
    required this.kind,
    this.name,
    this.type,
    this.className,
    this.returnType,
    this.sourceName,
    this.targetName,
    this.relationshipType,
    this.sourceMultiplicity,
    this.targetMultiplicity,
    this.label,
  });

  final VoiceCommandKind kind;
  final String? name;
  final String? type;
  final String? className;
  final String? returnType;
  final String? sourceName;
  final String? targetName;
  final VoiceRelationshipType? relationshipType;
  final String? sourceMultiplicity;
  final String? targetMultiplicity;
  final String? label;
}

const _classPattern = r'^crear\s+(?:una\s+)?clase\s+(.+)$';
const _relationshipPattern = r'^relacionar\s+(.+?)\s+con\s+(.+)$';
const _undoPattern =
    r'^deshacer(?:\s+(?:comando\s+de\s+voz|[úu]ltimo\s+comando\s+de\s+voz))?$';
const _attributePattern =
    r'^(?:agregar|a[ñn]adir)\s+atributo\s+(.+?)\s+de\s+tipo\s+(.+?)\s+a\s+(.+)$';
const _methodPattern =
    r'^(?:agregar|a[ñn]adir)\s+m[ée]todo\s+(.+?)\s+de\s+retorno\s+(.+?)\s+a\s+(.+)$';
final _namePattern = RegExp(r"^[\p{L}\p{N}][\p{L}\p{N}' -]*$", unicode: true);
final _multiplicityPattern = RegExp(
  r'^(?:\d+|\*)(?:\.\.(?:\d+|\*))?$',
);

// Punctuation Deepgram's smart_format inserts inside a transcript (commas,
// semicolons, colons, Spanish inverted marks, and straight/curly/angle
// quotes) that would otherwise break downstream patterns such as
// _cleanName.
final _innerPunctuationPattern = RegExp(r'[,;:¿¡"“”«»]');

/// Strips smart_format punctuation and collapses whitespace before parsing.
String _normalizeTranscript(String raw) {
  final withoutPunctuation = raw.replaceAll(_innerPunctuationPattern, ' ');
  final collapsed = withoutPunctuation.replaceAll(RegExp(r'\s+'), ' ').trim();
  return collapsed.replaceFirst(RegExp(r'[.!?]+$'), '').trim();
}

/// Maps the small Spanish attribute/return type vocabulary (case/accent
/// insensitive) a voice transcript can produce to the canonical Java type
/// name jdlgen expects. Unknown types (including already-canonical English
/// names such as "String") pass through unchanged.
String _mapSpanishType(String value) {
  switch (value.trim().toLowerCase()) {
    case 'entero':
      return 'Integer';
    case 'texto':
    case 'cadena':
      return 'String';
    case 'decimal':
      return 'BigDecimal';
    case 'fecha':
      return 'LocalDate';
    case 'fecha hora':
    case 'fecha y hora':
      return 'ZonedDateTime';
    case 'booleano':
    case 'lógico':
    case 'logico':
      return 'Boolean';
    case 'largo':
      return 'Long';
    case 'flotante':
      return 'Float';
    case 'doble':
      return 'Double';
    case 'uuid':
      return 'UUID';
    default:
      return value;
  }
}

VoiceCommand? parseVoiceCommand(String transcript) {
  final text = _normalizeTranscript(transcript);
  if (text.isEmpty ||
      RegExp(r'\bpar[aá]metros?\b', caseSensitive: false).hasMatch(text)) {
    return null;
  }
  if (RegExp(_undoPattern, caseSensitive: false).hasMatch(text)) {
    return const VoiceCommand(kind: VoiceCommandKind.undoVoiceCommand);
  }

  final attribute = RegExp(
    _attributePattern,
    caseSensitive: false,
  ).firstMatch(text);
  if (attribute != null) {
    final name = _cleanName(attribute.group(1)!);
    final type = _cleanName(attribute.group(2)!);
    final className = _cleanName(attribute.group(3)!);
    if (name == null || type == null || className == null) return null;
    return VoiceCommand(
      kind: VoiceCommandKind.addAttribute,
      name: name,
      type: _mapSpanishType(type),
      className: className,
    );
  }

  final method = RegExp(_methodPattern, caseSensitive: false).firstMatch(text);
  if (method != null) {
    final name = _cleanName(method.group(1)!);
    final returnType = _cleanName(method.group(2)!);
    final className = _cleanName(method.group(3)!);
    if (name == null || returnType == null || className == null) return null;
    return VoiceCommand(
      kind: VoiceCommandKind.addMethod,
      name: name,
      returnType: _mapSpanishType(returnType),
      className: className,
    );
  }

  final classMatch = RegExp(
    _classPattern,
    caseSensitive: false,
  ).firstMatch(text);
  if (classMatch != null) {
    final name = _cleanName(classMatch.group(1)!);
    return name == null
        ? null
        : VoiceCommand(kind: VoiceCommandKind.createClass, name: name);
  }

  return _parseRelationship(text);
}

VoiceCommand? _parseRelationship(String text) {
  final match = RegExp(
    _relationshipPattern,
    caseSensitive: false,
  ).firstMatch(text);
  if (match == null) return null;

  final sourceName = _cleanName(match.group(1)!);
  if (sourceName == null) return null;
  final rest = match.group(2)!.trim();
  final markerPattern = RegExp(
    r'\b(como|con\s+cardinalidad|con\s+verbo)\b',
    caseSensitive: false,
  );
  final markers = markerPattern.allMatches(rest).toList();
  final targetRaw =
      markers.isEmpty ? rest : rest.substring(0, markers.first.start);
  final targetName = _cleanName(targetRaw);
  if (targetName == null) return null;

  VoiceRelationshipType? type;
  String? sourceMultiplicity;
  String? targetMultiplicity;
  String? label;
  var hasType = false;
  var hasLabel = false;

  for (var index = 0; index < markers.length; index++) {
    final marker =
        markers[index].group(1)!.toLowerCase().replaceAll(RegExp(r'\s+'), ' ');
    final start = markers[index].end;
    final end =
        index + 1 < markers.length ? markers[index + 1].start : rest.length;
    final payload = rest.substring(start, end).trim();
    if (marker == 'como') {
      if (hasType) return null;
      hasType = true;
      type = _parseRelationshipType(payload);
      if (type == null) return null;
    } else if (marker == 'con cardinalidad') {
      final cardinalities = _parseCardinalities(payload);
      if (cardinalities == null) return null;
      if (cardinalities.$1 != null && sourceMultiplicity != null ||
          cardinalities.$2 != null && targetMultiplicity != null) {
        return null;
      }
      sourceMultiplicity ??= cardinalities.$1;
      targetMultiplicity ??= cardinalities.$2;
    } else if (marker == 'con verbo') {
      if (hasLabel) return null;
      hasLabel = true;
      label = _cleanName(payload);
      if (label == null) return null;
    } else {
      return null;
    }
  }

  return VoiceCommand(
    kind: VoiceCommandKind.createRelationship,
    sourceName: sourceName,
    targetName: targetName,
    relationshipType: type,
    sourceMultiplicity: sourceMultiplicity,
    targetMultiplicity: targetMultiplicity,
    label: label,
  );
}

VoiceRelationshipType? _parseRelationshipType(String value) {
  switch (value.trim().toLowerCase()) {
    case 'asociación':
    case 'asociacion':
    case 'association':
      return VoiceRelationshipType.association;
    case 'agregación':
    case 'agregacion':
    case 'aggregation':
      return VoiceRelationshipType.aggregation;
    case 'composición':
    case 'composicion':
    case 'composition':
      return VoiceRelationshipType.composition;
    case 'generalización':
    case 'generalizacion':
    case 'generalization':
      return VoiceRelationshipType.generalization;
    case 'realización':
    case 'realizacion':
    case 'realization':
      return VoiceRelationshipType.realization;
    case 'dependencia':
    case 'dependency':
      return VoiceRelationshipType.dependency;
    default:
      return null;
  }
}

(String?, String?)? _parseCardinalities(String payload) {
  final bothSourceTarget = RegExp(
    r'^origen\s+(.+?)\s+y\s+(?:cardinalidad\s+)?destino\s+(.+)$',
    caseSensitive: false,
  ).firstMatch(payload);
  if (bothSourceTarget != null) {
    final source = _cleanMultiplicity(bothSourceTarget.group(1)!);
    final target = _cleanMultiplicity(bothSourceTarget.group(2)!);
    return source == null || target == null ? null : (source, target);
  }
  final bothTargetSource = RegExp(
    r'^destino\s+(.+?)\s+y\s+(?:cardinalidad\s+)?origen\s+(.+)$',
    caseSensitive: false,
  ).firstMatch(payload);
  if (bothTargetSource != null) {
    final target = _cleanMultiplicity(bothTargetSource.group(1)!);
    final source = _cleanMultiplicity(bothTargetSource.group(2)!);
    return source == null || target == null ? null : (source, target);
  }
  final sourceOnly = RegExp(
    r'^origen\s+(.+)$',
    caseSensitive: false,
  ).firstMatch(payload);
  if (sourceOnly != null) {
    final source = _cleanMultiplicity(sourceOnly.group(1)!);
    return source == null ? null : (source, null);
  }
  final targetOnly = RegExp(
    r'^destino\s+(.+)$',
    caseSensitive: false,
  ).firstMatch(payload);
  if (targetOnly != null) {
    final target = _cleanMultiplicity(targetOnly.group(1)!);
    return target == null ? null : (null, target);
  }
  return null;
}

String? _cleanName(String value) {
  final name = value.trim().replaceAll(RegExp(r'\s+'), ' ');
  return name.isNotEmpty && name.length <= 80 && _namePattern.hasMatch(name)
      ? name
      : null;
}

// _spokenNumbers maps the small Spanish cardinal vocabulary a speech-to-text
// transcript can produce for a UML multiplicity bound to its digit form.
const _spokenNumbers = {
  'cero': '0',
  'uno': '1',
  'dos': '2',
  'tres': '3',
  'cuatro': '4',
  'cinco': '5',
  'seis': '6',
  'siete': '7',
  'ocho': '8',
  'nueve': '9',
  'diez': '10',
};

final _digitsOnly = RegExp(r'^\d+$');

/// Resolves one spoken or literal multiplicity bound token ("uno", "5",
/// "muchos", "*") to its digit/"*" form, or null when unrecognized.
String? _spokenBound(String token) {
  final norm = token.trim().toLowerCase();
  if (norm == '*' ||
      norm == 'muchos' ||
      norm == 'varios' ||
      norm == 'asterisco' ||
      norm == 'estrella' ||
      norm == 'n') {
    return '*';
  }
  if (_digitsOnly.hasMatch(norm)) return norm;
  return _spokenNumbers[norm];
}

String? _checkRange(String value) {
  return _multiplicityPattern.hasMatch(value) ? value : null;
}

/// Parses one multiplicity payload into _multiplicityPattern form. Accepts
/// the literal digit/"*" forms plus the spoken Spanish vocabulary a Deepgram
/// transcript produces: a single value ("uno", "muchos"), "X a Y", "de X a
/// Y", "X punto punto Y", the literal "X..Y", "X o más"/"X o mas" ("X..*"),
/// and the idiom "cero o uno" ("0..1").
String? _cleanMultiplicity(String value) {
  final trimmed = value.trim().replaceAll(RegExp(r'\s*\.\.\s*'), '..');
  if (_multiplicityPattern.hasMatch(trimmed)) return trimmed;

  final lower = trimmed.toLowerCase();

  if (RegExp(r'^cero\s+o\s+uno$').hasMatch(lower)) return '0..1';

  final orMore = RegExp(r'^(.+?)\s+o\s+m[aá]s$').firstMatch(lower);
  if (orMore != null) {
    final lo = _spokenBound(orMore.group(1)!);
    return lo == null ? null : _checkRange('$lo..*');
  }

  final deXaY = RegExp(r'^de\s+(.+?)\s+a\s+(.+)$').firstMatch(lower);
  if (deXaY != null) {
    final lo = _spokenBound(deXaY.group(1)!);
    final hi = _spokenBound(deXaY.group(2)!);
    return lo == null || hi == null ? null : _checkRange('$lo..$hi');
  }

  final puntoPunto =
      RegExp(r'^(.+?)\s+punto\s+punto\s+(.+)$').firstMatch(lower);
  if (puntoPunto != null) {
    final lo = _spokenBound(puntoPunto.group(1)!);
    final hi = _spokenBound(puntoPunto.group(2)!);
    return lo == null || hi == null ? null : _checkRange('$lo..$hi');
  }

  final xaY = RegExp(r'^(.+?)\s+a\s+(.+)$').firstMatch(lower);
  if (xaY != null) {
    final lo = _spokenBound(xaY.group(1)!);
    final hi = _spokenBound(xaY.group(2)!);
    return lo == null || hi == null ? null : _checkRange('$lo..$hi');
  }

  final single = _spokenBound(lower);
  return single == null ? null : _checkRange(single);
}

File: frontend/src/diagram/voiceCommands.ts
import { RelationshipType } from '../types';
import { MULTIPLICITY_PATTERN } from './relationshipHelpers';

export type VoiceCommand =
  | { kind: 'create-class'; name: string }
  | {
      kind: 'create-relationship';
      sourceName: string;
      targetName: string;
      type?: RelationshipType;
      sourceMultiplicity?: string;
      targetMultiplicity?: string;
      label?: string;
    }
  | { kind: 'undo-voice-command' }
  | { kind: 'add-attribute'; name: string; type: string; className: string }
  | { kind: 'add-method'; name: string; returnType: string; className: string };

const CLASS_PATTERN = /^crear\s+(?:una\s+)?clase\s+(.+)$/i;
const RELATIONSHIP_START_PATTERN = /^relacionar\s+(.+?)\s+con\s+(.+)$/i;
const UNDO_PATTERN = /^deshacer(?:\s+(?:comando\s+de\s+voz|último\s+comando\s+de\s+voz))?$/i;
const ATTRIBUTE_PATTERN = /^(?:agregar|añadir)\s+atributo\s+(.+?)\s+de\s+tipo\s+(.+?)\s+a\s+(.+)$/i;
const METHOD_PATTERN = /^(?:agregar|añadir)\s+método\s+(.+?)\s+de\s+retorno\s+(.+?)\s+a\s+(.+)$/i;
const NAME_PATTERN = /^[\p{L}\p{N}][\p{L}\p{N}' -]*$/u;

function cleanName(value: string): string | null {
  const name = value.trim().replace(/\s+/g, ' ');
  return name.length > 0 && name.length <= 80 && NAME_PATTERN.test(name) ? name : null;
}

function cleanLabel(value: string): string | null {
  const label = value.trim().replace(/\s+/g, ' ');
  return label.length > 0 && label.length <= 80 && NAME_PATTERN.test(label) ? label : null;
}

function parseRelationshipType(value: string): RelationshipType | null {
  const norm = value.trim().toLowerCase();
  if (/^(?:asociaci[oó]n|association)$/.test(norm)) return 'association';
  if (/^(?:agregaci[oó]n|aggregation)$/.test(norm)) return 'aggregation';
  if (/^(?:composici[oó]n|composition)$/.test(norm)) return 'composition';
  if (/^(?:generalizaci[oó]n|generalization)$/.test(norm)) return 'generalization';
  if (/^(?:realizaci[oó]n|realization)$/.test(norm)) return 'realization';
  if (/^(?:dependencia|dependency)$/.test(norm)) return 'dependency';
  return null;
}

// SPOKEN_NUMBERS maps the small Spanish cardinal vocabulary a speech-to-text
// transcript can produce for a UML multiplicity bound to its digit form.
const SPOKEN_NUMBERS: Record<string, string> = {
  cero: '0',
  uno: '1',
  dos: '2',
  tres: '3',
  cuatro: '4',
  cinco: '5',
  seis: '6',
  siete: '7',
  ocho: '8',
  nueve: '9',
  diez: '10',
};

/** Resolves one spoken or literal multiplicity bound token ("uno", "5", "muchos", "*") to its digit/"*" form, or null when unrecognized. */
function spokenBound(token: string): string | null {
  const norm = token.trim().toLowerCase();
  if (norm === '*' || norm === 'muchos' || norm === 'varios' || norm === 'asterisco' || norm === 'estrella' || norm === 'n') {
    return '*';
  }
  if (/^\d+$/.test(norm)) return norm;
  return SPOKEN_NUMBERS[norm] ?? null;
}

/**
 * Parses one multiplicity payload into MULTIPLICITY_PATTERN form. Accepts the
 * literal digit/"*" forms plus the spoken Spanish vocabulary a Deepgram
 * transcript produces: a single value ("uno", "muchos"), "X a Y", "de X a Y",
 * "X punto punto Y", the literal "X..Y", "X o más"/"X o mas" ("X..*"), and the
 * idiom "cero o uno" ("0..1").
 */
function cleanMultiplicity(value: string): string | null {
  const trimmed = value.trim().replace(/\s*\.\.\s*/g, '..');
  if (MULTIPLICITY_PATTERN.test(trimmed)) return trimmed;

  const lower = trimmed.toLowerCase();

  if (/^cero\s+o\s+uno$/.test(lower)) return '0..1';

  const orMore = /^(.+?)\s+o\s+m[aá]s$/.exec(lower);
  if (orMore) {
    const lo = spokenBound(orMore[1]);
    return lo ? checkRange(`${lo}..*`) : null;
  }

  const deXaY = /^de\s+(.+?)\s+a\s+(.+)$/.exec(lower);
  if (deXaY) {
    const lo = spokenBound(deXaY[1]);
    const hi = spokenBound(deXaY[2]);
    return lo && hi ? checkRange(`${lo}..${hi}`) : null;
  }

  const puntoPunto = /^(.+?)\s+punto\s+punto\s+(.+)$/.exec(lower);
  if (puntoPunto) {
    const lo = spokenBound(puntoPunto[1]);
    const hi = spokenBound(puntoPunto[2]);
    return lo && hi ? checkRange(`${lo}..${hi}`) : null;
  }

  const xaY = /^(.+?)\s+a\s+(.+)$/.exec(lower);
  if (xaY) {
    const lo = spokenBound(xaY[1]);
    const hi = spokenBound(xaY[2]);
    return lo && hi ? checkRange(`${lo}..${hi}`) : null;
  }

  const single = spokenBound(lower);
  return single ? checkRange(single) : null;
}

function checkRange(value: string): string | null {
  return MULTIPLICITY_PATTERN.test(value) ? value : null;
}

function parseCardinalities(payload: string): { sourceMultiplicity?: string; targetMultiplicity?: string } | null {
  const norm = payload.trim();
  const bothSrcDst = /^origen\s+(.+?)\s+y\s+(?:cardinalidad\s+)?destino\s+(.+)$/i.exec(norm);
  if (bothSrcDst) {
    const src = cleanMultiplicity(bothSrcDst[1]);
    const dst = cleanMultiplicity(bothSrcDst[2]);
    return src && dst ? { sourceMultiplicity: src, targetMultiplicity: dst } : null;
  }
  const bothDstSrc = /^destino\s+(.+?)\s+y\s+(?:cardinalidad\s+)?origen\s+(.+)$/i.exec(norm);
  if (bothDstSrc) {
    const dst = cleanMultiplicity(bothDstSrc[1]);
    const src = cleanMultiplicity(bothDstSrc[2]);
    return src && dst ? { sourceMultiplicity: src, targetMultiplicity: dst } : null;
  }
  const srcOnly = /^origen\s+(.+)$/i.exec(norm);
  if (srcOnly) {
    const src = cleanMultiplicity(srcOnly[1]);
    return src ? { sourceMultiplicity: src } : null;
  }
  const dstOnly = /^destino\s+(.+)$/i.exec(norm);
  if (dstOnly) {
    const dst = cleanMultiplicity(dstOnly[1]);
    return dst ? { targetMultiplicity: dst } : null;
  }
  return null;
}

function parseRelationshipCommand(text: string): VoiceCommand | null {
  const match = RELATIONSHIP_START_PATTERN.exec(text);
  if (!match) return null;

  const sourceName = cleanName(match[1]);
  if (!sourceName) return null;

  const rest = match[2].trim();
  const markerRegex = /\b(como|con\s+cardinalidad|con\s+verbo)\b/gi;
  const markers: Array<{ marker: string; index: number; length: number }> = [];
  let m: RegExpExecArray | null;
  while ((m = markerRegex.exec(rest)) !== null) {
    markers.push({
      marker: m[1].toLowerCase().replace(/\s+/g, ' '),
      index: m.index,
      length: m[0].length,
    });
  }

  let targetRaw: string;
  const modifierTokens: Array<{ marker: string; payload: string }> = [];

  if (markers.length === 0) {
    targetRaw = rest;
  } else {
    targetRaw = rest.slice(0, markers[0].index);
    for (let i = 0; i < markers.length; i++) {
      const current = markers[i];
      const start = current.index + current.length;
      const end = i + 1 < markers.length ? markers[i + 1].index : rest.length;
      const payload = rest.slice(start, end).trim();
      modifierTokens.push({ marker: current.marker, payload });
    }
  }

  const targetName = cleanName(targetRaw);
  if (!targetName) return null;

  let relationshipType: RelationshipType | undefined;
  let sourceMultiplicity: string | undefined;
  let targetMultiplicity: string | undefined;
  let label: string | undefined;

  let hasType = false;
  let hasLabel = false;

  for (const token of modifierTokens) {
    if (token.marker === 'como') {
      if (hasType) return null;
      hasType = true;
      const parsedType = parseRelationshipType(token.payload);
      if (!parsedType) return null;
      relationshipType = parsedType;
    } else if (token.marker === 'con cardinalidad') {
      const cards = parseCardinalities(token.payload);
      if (!cards) return null;
      if (cards.sourceMultiplicity !== undefined) {
        if (sourceMultiplicity !== undefined) return null;
        sourceMultiplicity = cards.sourceMultiplicity;
      }
      if (cards.targetMultiplicity !== undefined) {
        if (targetMultiplicity !== undefined) return null;
        targetMultiplicity = cards.targetMultiplicity;
      }
    } else if (token.marker === 'con verbo') {
      if (hasLabel) return null;
      hasLabel = true;
      const cleanedLabel = cleanLabel(token.payload);
      if (!cleanedLabel) return null;
      label = cleanedLabel;
    } else {
      return null;
    }
  }

  return {
    kind: 'create-relationship',
    sourceName,
    targetName,
    ...(relationshipType !== undefined ? { type: relationshipType } : {}),
    ...(sourceMultiplicity !== undefined ? { sourceMultiplicity } : {}),
    ...(targetMultiplicity !== undefined ? { targetMultiplicity } : {}),
    ...(label !== undefined ? { label } : {}),
  };
}

// INNER_PUNCTUATION_PATTERN matches the punctuation Deepgram's smart_format
// inserts inside a transcript (commas, semicolons, colons, Spanish inverted
// marks, and straight/curly/angle quotes) that would otherwise break
// downstream patterns such as cleanName.
const INNER_PUNCTUATION_PATTERN = /[,;:¿¡"“”«»]/g;

/** Strips smart_format punctuation and collapses whitespace before parsing. */
function normalizeTranscript(raw: string): string {
  const withoutPunctuation = raw.replace(INNER_PUNCTUATION_PATTERN, ' ');
  const collapsed = withoutPunctuation.replace(/\s+/g, ' ').trim();
  return collapsed.replace(/[.!?]+$/g, '').trim();
}

/**
 * Maps the small Spanish attribute/return type vocabulary (case/accent
 * insensitive) a voice transcript can produce to the canonical Java type
 * name jdlgen expects. Unknown types (including already-canonical English
 * names such as "String") pass through unchanged.
 */
function mapSpanishType(value: string): string {
  switch (value.trim().toLowerCase()) {
    case 'entero':
      return 'Integer';
    case 'texto':
    case 'cadena':
      return 'String';
    case 'decimal':
      return 'BigDecimal';
    case 'fecha':
      return 'LocalDate';
    case 'fecha hora':
    case 'fecha y hora':
      return 'ZonedDateTime';
    case 'booleano':
    case 'lógico':
    case 'logico':
      return 'Boolean';
    case 'largo':
      return 'Long';
    case 'flotante':
      return 'Float';
    case 'doble':
      return 'Double';
    case 'uuid':
      return 'UUID';
    default:
      return value;
  }
}

/** Parses only the small, confirmed Spanish voice vocabulary. */
export function parseVoiceCommand(transcript: string): VoiceCommand | null {
  const text = normalizeTranscript(transcript);
  // Parameters are intentionally outside this first member-command vocabulary.
  if (/\bpar[aá]metros?\b/i.test(text)) return null;
  if (UNDO_PATTERN.test(text)) return { kind: 'undo-voice-command' };

  const attributeMatch = ATTRIBUTE_PATTERN.exec(text);
  if (attributeMatch) {
    const name = cleanName(attributeMatch[1]);
    const type = cleanName(attributeMatch[2]);
    const className = cleanName(attributeMatch[3]);
    return name && type && className
      ? { kind: 'add-attribute', name, type: mapSpanishType(type), className }
      : null;
  }

  const methodMatch = METHOD_PATTERN.exec(text);
  if (methodMatch) {
    const name = cleanName(methodMatch[1]);
    const returnType = cleanName(methodMatch[2]);
    const className = cleanName(methodMatch[3]);
    return name && returnType && className
      ? { kind: 'add-method', name, returnType: mapSpanishType(returnType), className }
      : null;
  }

  const classMatch = CLASS_PATTERN.exec(text);
  if (classMatch) {
    const name = cleanName(classMatch[1]);
    return name ? { kind: 'create-class', name } : null;
  }

  return parseRelationshipCommand(text);
}
